pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+	shouldCache queryrangebase.ShouldCacheFn
+
+	metrics *LogResultCacheMetrics
+	logger  log.Logger
+}
+
+func (l *logResultCache) Do(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	if l.shouldCache != nil && !l.shouldCache(req) {
+		return l.next.Do(ctx, req)
+	}
+
+	maxCacheFreshness := validation.MaxDurationPerTenant(tenantIDs, l.limits.MaxCacheFreshness)
+	maxCacheTime := int64(model.Now().Add(-maxCacheFreshness))
+	if req.GetStart() > maxCacheTime {
+		return l.next.Do(ctx, req)
+	}
+
+	lokiReq, ok := req.(*LokiRequest)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "invalid request type %T", req)
+	}
+
+	interval := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, l.limits.QuerySplitDuration)
+	// skip caching by if interval is unset
+	if interval == 0 {
+		return l.next.Do(ctx, req)
+	}
+	// The first subquery might not be aligned.
+	alignedStart := time.Unix(0, lokiReq.GetStartTs().UnixNano()-(lokiReq.GetStartTs().UnixNano()%interval.Nanoseconds()))
+	// generate the cache key based on query, tenant and start time.
+	cacheKey := fmt.Sprintf("log:%s:%s:%d", tenant.JoinTenantIDs(tenantIDs), req.GetQuery(), alignedStart.UnixNano()/(interval.Nanoseconds()))
+
+	_, buff, _, err := l.cache.Fetch(ctx, []string{cache.HashKey(cacheKey)})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error fetching cache", "err", err, "cacheKey", cacheKey)
+		return l.next.Do(ctx, req)
+	}
+	// we expect only one key to be found or missing.
+	if len(buff) > 1 {
+		level.Warn(l.logger).Log("msg", "unexpected length of cache return values", "buff", len(buff))
+		return l.next.Do(ctx, req)
+	}
+
+	if len(buff) == 0 {
+		// cache miss
+		return l.handleMiss(ctx, cacheKey, lokiReq)
+	}
+
+	// cache hit
+	var cachedResquest LokiRequest
+	err = proto.Unmarshal(buff[0], &cachedResquest)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error unmarshalling request from cache", "err", err)
+		return l.next.Do(ctx, req)
+	}
+	return l.handleHit(ctx, cacheKey, &cachedResquest, lokiReq)
+}
+
+func (l *logResultCache) handleMiss(ctx context.Context, cacheKey string, req *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheMiss.Inc()
+	level.Debug(l.logger).Log("msg", "cache miss", "key", cacheKey)
+	resp, err := l.next.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	lokiRes, ok := resp.(*LokiResponse)
+	if !ok {
+		return nil, fmt.Errorf("unexpected response type %T", resp)
+	}
+	// At the moment we only cache empty results
+	if !isEmpty(lokiRes) {
+		return resp, nil
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error marshalling request", "err", err)
+		return resp, nil
+	}
+	// cache the result
+	err = l.cache.Store(ctx, []string{cache.HashKey(cacheKey)}, [][]byte{data})
```

Comment:
I think this should be done in the background to avoid introducing extra latency, especially if we're seeing cache-related issues.

Reply:
I don't think caching small item like this will be a problem, specially that the timeout is super aggressive. I don't feel like doing this just yet unless it shows to be a problem in the future.